first commit

golnazads opened the pull request.

golnazads committed:
first commit

golnazads committed:
to start a build

golnazads committed:
Merge branch 'main' into main

golnazads committed:
added couple of comments to add more commands for easier use

golnazads committed:
fixed yml

golnazads committed:
Merge branch 'main' into main

golnazads committed:
Update python_actions.yml

golnazads committed:
Update python_actions.yml

golnazads committed:
Update python_actions.yml

golnazads committed:
Update python_actions.yml

golnazads committed:
Merge branch 'main' into main

golnazads committed:
rewrote a unittest that fails upstream, but passes locally, hopefully it will pass now

golnazads committed:
Merge remote-tracking branch 'origin/main'

golnazads committed:
completed the readme and fixed a type in the commeand of run.py

kelockhart commented on run.py:
```diff
@@ -0,0 +1,147 @@
+    parser.add_argument('-a', '--action', help='one of the followings: collect, identify, end_to_end (collect followed by identify).')
+    parser.add_argument('-t', '--target', help='target (ie, Moon, Mars, etc) is required. Has to be capitalized.')
+    parser.add_argument('-f', '--feature_type', help='feature type (ie, Crater or Albedo Feature), is one of the options, either feature type or feature name are required, but no both. Has to be capitalized and in singular form.')
+    parser.add_argument('-n', '--feature_name', help='feature name (ie Apollo, Atlas), is one of the options, either feature type or feature name are required, but no both. Has to be capitalized.')
```
Is there a full list of available options for `target`, `feature_type`, and `feature_name`?

golnazads replied:
yes, in database, and init in alembic.

golnazads replied:
regarding second bullet point, yes, I would like to deploy to dev and you guys tested it. this needs lots of resources. I did some ablation testing to see what I can eliminate, and I was not to eliminate any of the libraries.
I like to populate the database, collect data for knowledge graph, this is a one time thing per feature name, then it can be tested for identifying unknown entities, which is a regular operation.

golnazads replied:
OK, so moving the password for the Brain to local_config.
That should be it for this so closing it as done.

kelockhart commented on run.py:
```diff
@@ -0,0 +1,147 @@
+                    elif action_type == PLANETARYNAMES_PIPELINE_ACTION.remove_the_most_recent:
+                        KBH_rows_deleted, KB_rows_deleted = app.remove_most_recent_knowledge_base_records(feature_name_entity=feature_name,
+                                                                                                          target_entity=current_target)
+                        print(f"Removed the most recent records of knowledge graph for feature name `{feature_name}` and target `{current_target}`.\n"
```
Let's add log statements (or replace the print statements with logs) for all of the print statements in this file

golnazads replied:
OK. I am guessing that is how we want to proceed from here on, right?
Make a note to do the same for reference pipeline.

kelockhart replied:
Yup, once we have it running, it'll be easier to have log statements throughout.

golnazads replied:
done. closing the comment.

kelockhart commented on adsplanetnamepipe/app.py:
```diff
@@ -0,0 +1,437 @@
+        """
+        return all the feature names for target and feature type
+
+        :param target_entity:
```
Fill out all the docstrings in here

golnazads replied:
Not getting it. Elaborate please.

kelockhart replied:
I think I meant this to be a general comment - not all of the functions have a function description (like this one does) in the docstring, and many of them are missing descriptions of the params in the docstrings. More documentation on what the functions do, plus what the expected input params are, would be helpful

golnazads replied:
the function names are descriptive, or i thought so.

kelockhart commented on docs/database_UML.png:
These diagrams are great - makes it much easier to see how things flow.

golnazads replied:
I am glad. Not the database but the code diagram helped me revise the code for better.

kelockhart commented on adsplanetnamepipe/app.py:
```diff
@@ -0,0 +1,437 @@
+        with self.session_scope() as session:
+            rows = session.query(FeatureName).filter(and_(FeatureName.target_entity ==  target_entity,
+                                                          FeatureName.entity == feature_name_entity)).all()
+            if len(rows) == 1:
```
Is it possible to have ambiguity here, e.g. two different types of features on a given body with the same name, so that `len(rows)` is legitimately > 1?

golnazads replied:
No. feature names are ambiguous but at least did not not go name different feature types for the same target (body), lol. That would have been challenging, lol.

kelockhart commented on adsplanetnamepipe/app.py:
```diff
@@ -0,0 +1,437 @@
+        """
+        with self.session_scope() as session:
+            rows = session.query(AmbiguousFeatureName).filter(AmbiguousFeatureName.entity == feature_name_entity).all()
+            if len(rows) > 1:
```
I'm guessing you actually want this to be `>=` (here and in the next few handler functions)

golnazads replied:
fixed.

kelockhart commented on adsplanetnamepipe/utils/search_retrieval.py:
```diff
@@ -0,0 +1,174 @@
+        "SSRv"
+    ])
+
+    other_filters = 'database:astronomy fulltext_mtime:["2000-01-01t00:00:00.000Z" TO *]'
```
Consider making all of this editable in the config - eventually we'll want to process the full pre-2000 back catalog, plus eventually/soon we'll also have a planetary science database

golnazads replied:
good call. I shall.

golnazads replied:
done.

kelockhart commented on adsplanetnamepipe/utils/search_retrieval.py:
```diff
@@ -0,0 +1,174 @@
+        :param query:
+        :return:
+        """
+        index = 1
```
Solr is 0-indexed, so start this at 0 instead so you don't miss the first row

golnazads replied:
Wow, I know this, I should know this, lol. Great catch.

kelockhart commented on adsplanetnamepipe/utils/search_retrieval.py:
```diff
@@ -0,0 +1,174 @@
+
+        :return:
+        """
+        query = f'full:(="{self.args.feature_name}") full:("{self.args.target}") full:("{self.feature_types_ored}") year:[2000 TO *]'
```
Same comment as above - we'll eventually need to process records before 2000, so this should be configurable, either in the config or at run time

golnazads replied:
done.

kelockhart commented on adsplanetnamepipe/utils/search_retrieval.py:
```diff
@@ -0,0 +1,174 @@
+        :return:
+        """
+        query = f'full:(="{self.args.feature_name}") full:("{self.args.target}") full:("{self.feature_types_ored}") year:[2000 TO *]'
+        query += self.planetry_journal_filter + self.other_filters
```
I see your note on `self.planetry_journal_filter` above that Mike specified these journals only for now, but we'll presumably expand beyond this set eventually, so consider making this configurable as well.

golnazads replied:
done.

kelockhart commented on adsplanetnamepipe/utils/search_retrieval.py:
```diff
@@ -0,0 +1,174 @@
+        query += f'-{self.other_filters}'
+
+        # 200 is enough for the negative side
+        docs, status_code = self.single_solr_query(query=query, start=1, rows=500)
```
`start=0` to avoid missing the first record

golnazads replied:
done

kelockhart commented on adsplanetnamepipe/utils/common.py:
```diff
@@ -0,0 +1,150 @@
+class Synonyms(object):
+
+    # generated 5/20/2022
+    synonym_list = {
```
How frequently do you see this being updated? If not very often, this is fine; otherwise consider moving to a config

golnazads replied:
Need to ask, I dont know the answer.

golnazads replied:
I asked Alberto about this, and he said that he talked to you and Taylor. I brought up the same issue on the side of resolver with Taylor. Over there, I am one step closer. There are files on the server that Edwin modifies on a regular basis. They are called source matchers. Reading and processing them takes some time, so I do that offline, create a serializable object, and then load that every time the service is deployed. I actually went through the motion, got files from the server, created a serializable object, uploaded it to GitHub, and asked Taylor to see if he could upload the serializable object to the pod without making a release. It was just before his vacation. He said he would think about it afterward. So, I think I should do the same thing here as well: make a serializable object and update it on a regular basis. I will make an issue out of this and wait to see if it works for the resolver. If it does, then I’ll proceed this way; otherwise, I’ll go with another approach.

created the issue https://github.com/adsabs/ADSPlanetaryNamesPipeline/issues/2

kelockhart commented on adsplanetnamepipe/utils/match_excerpt.py:
```diff
@@ -0,0 +1,240 @@
+        if the_sum > 0:
+            # normalize
+            ambiguous_context_count = [float(term) / the_sum for term in ambiguous_context_count]
+            # sort both the terms and values ascending
```
I think you mean `descending` here (at least that's what the code is doing, and it makes sense)

golnazads replied:
yes, lol.

kelockhart commented on adsplanetnamepipe/utils/match_excerpt.py:
```diff
@@ -0,0 +1,240 @@
+        """
+
+        valid = self.spacy.validate_feature_name(text, self.args, entity_span, usgs_term) and \
+                self.yake.validate_feature_name(text, self.args, entity_span, usgs_term)
```
Out of curiosity, why validate with both the Spacy and the Yake models?

golnazads replied:
This is to make sure that the feature name is not part of the phrase. They identify phrases differently. As you know spacy does not have any tuning parameter, but yake does. I tried to change the yake parameter to identify the same phrases as spacy to no luck, and hence both are being used. But it is important that we catch feature name being part of a phrase that is indication that this is not planetary name.

kelockhart commented on adsplanetnamepipe/collect.py:
```diff
@@ -0,0 +1,186 @@
+        text = ' '.join(doc.get('title', '')) + ' ' + doc.get('abstract', '') + ' ' + doc.get('body', '')
+        return self.paper_relevance.forward(
+            text,
+            bibstem=doc.get('bibcode', '')[4:9].strip('.'),
```
To consider for the future: we'll eventually stop using bibcodes for every record, so this will eventually need to be updated to retrieve the journal in another manner

golnazads replied:
Yes, not sure how do we want to proceed here. Need to chat about this further.

kelockhart replied:
I'd leave it as-is for now, since we won't be moving away from bibcodes for a bit longer yet, but maybe make an issue on the repo to handle it later.

golnazads replied:
I also brought this up and i was told I should not be concerned, since probably this issue will not be solved anytime soon, and I probably retire before needed to be dealing with this, lol.

golnazads replied:
I also brought this up and was told that I shouldn’t be concerned, as this issue likely won’t be resolved anytime soon, and I’ll probably retire before needing to deal with it, lol.

kelockhart commented on adsplanetnamepipe/utils/extract_keywords.py:
```diff
@@ -0,0 +1,476 @@
+
+class WikiWrapper():
+
+    wiki_vocab = '''
```
This is hard to read/maintain in this format - can it get moved to a separate file? Line breaks would be helpful too - Github will have a hard time doing a diff on this if it's all one line

golnazads replied:
I am so determine to get rid of this, lol. But for now going to put it in a dat file. A group visited us and gave a presentation on planetary data, I asked them for their data hoping I could replace this, and they send it to me, turn out they are using the same wikidata structure. I am hoping that you guys consider hosting the Nasa Concept library. If you do, I can remove this.

golnazads replied:
done.

kelockhart commented on adsplanetnamepipe/utils/local_llm.py:
```diff
@@ -0,0 +1,50 @@
+        """
+        if abstract:
+            url = 'https://playground.adsabs.harvard.edu/brain/v1/chat'
+            headers = {'Content-Type': 'application/json', 'Authorization': 'Bearer %s'%'edge!'}
```
This needs to be in local_config

golnazads replied:
Yes.

golnazads replied:
done.

kelockhart commented on adsplanetnamepipe/tasks.py:
```diff
@@ -0,0 +1,70 @@
+            keywords_positive = app.get_knowledge_base_keywords(the_task['args'].feature_name,
+                                                               the_task['args'].feature_type,
+                                                               the_task['args'].target,
+                                                               the_task['args'].name_entity_labels[0]['label'])
```
This is assuming that `app.get_named_entity_label` is always going to return in the appropriate order, which feels easy for a future developer to screw up. I'd add an explicit check on `name_entity_labels['value']` instead

golnazads replied:
The named_entity_label table is a small lookup table with only two rows: 'planetary' and 'non planetary'. During the creation of the database, this table is initialized with 'planetary' inserted first. Since the query doesn't specify an order, it's highly likely that 'planetary' would already be the first item returned.
To make it foolproof, I implemented a sorting function in the NamedEntityLabel model and modified the get_named_entity_label method to use this sorting function after querying the database. This way, it is guaranteed that 'planetary' will always be the first item in the returned list.

kelockhart commented on adsplanetnamepipe/tasks.py:
```diff
@@ -0,0 +1,70 @@
+                return app.insert_knowledge_base_records(knowledge_base_records)
+            return False
+        # action to identify and label entities
+        if the_task['action_type'] in [PLANETARYNAMES_PIPELINE_ACTION.identify, PLANETARYNAMES_PIPELINE_ACTION.end_to_end]:
```
`PLANETARYNAMES_PIPELINE_ACTION.end_to_end` won't make it here - it'll get kicked back on a return on the `if` statement above

golnazads replied:
Really? Not sure if I had changed anything at the last minute that is not working for you. This is the step to predict the feature name. So if I am understanding you correctly, you are saying it does not go in the if statement, right?

kelockhart replied:
I didn't actually run it, but I saw the `return` in the previous if statement and assumed it would skip this one.

golnazads replied:
It is actually either/or thing:
either
comment: action to collect data to setup KB graph
or
comment: action to identify and label entities
and when it is done with the task it should leave.
I actually had thought of having two queues, one for collect, one for identify. Then I remembered how against having multiple queues Roman was, so I went with this.
Two options:
1- add comment specifying that the actions are either/or and as soon as the processing of the task is done leave
2- have two queues one for collect, one for identify. My vote is for this. My thinking is that each queue should contain a specific task. But Roman's philosophy, I think, was one pipeline/one queue. It did not make sense to me then, and it does not make sense to me now. I dont remember if he ever clearly explained his why!
Anyhow, take your pick, both approaches will take only a few minute to address.

golnazads replied:
I did not hear back from you, so I am going the route of adding more comments.

golnazads replied:
added comments.

kelockhart commented on adsplanetnamepipe/app.py:
```diff
@@ -0,0 +1,437 @@
+            rows = session.query(KnowledgeBaseHistory) \
+                .filter(and_(KnowledgeBaseHistory.feature_name_entity == feature_name_entity,
+                             KnowledgeBaseHistory.target_entity == target_entity,
+                             KnowledgeBaseHistory.named_entity_label == 'planetary')) \
```
Consider using the lookup table here, since you're using one

golnazads replied:
Not follow. Please elaborate.

kelockhart replied:
I was thinking along the lines of doing a join on the `NamedEntityLabel` table, on `KnowledgeBaseHistory.named_entity_label == NamedEntityLabel.label`

golnazads replied:
The join with the NamedEntityLabel table is not necessary for this query. Since there's a foreign key constraint between KnowledgeBaseHistory.named_entity_label and NamedEntityLabel.label, we don't need to explicitly join these tables for this operation. The foreign key ensures data integrity at the database level. Also since we're only filtering on the label value, which is directly stored in KnowledgeBaseHistory, we can use the simpler query without the join.

kelockhart commented on adsplanetnamepipe/utils/knowledge_graph.py:
```diff
@@ -0,0 +1,133 @@
+        pair_counts = defaultdict(int)
+
+        for keywords in keywords_list:
+            for keyword in keywords:
```
This will end up double counting pairs - not sure from the context if that's expected or not

golnazads replied:
No, should not. I will check it out. Note that there are two lists
https://github.com/golnazads/ADSPlanetaryNamesPipeline/blob/main/adsplanetnamepipe/tests/unittests/stubdata/keywords.py#L2
one keywords from spacy/yake/wikidata and one phrases from nasa concept, eventually.

golnazads replied:
checked this again, it does not double count in error. If you mean it counts both 'entropy' and 'diagnosis', also 'diagnosis' and 'entropy', that is by design since we want to count co occurrence in both direction.

kelockhart reviewed (commented):
Looks very good! There's a LOT of code in here (you've obviously invested a ton of effort) but it's really well organized and easy to follow. Most of my comments are about readability/maintainability, though there are a couple small typos that I point out here and there. Also, important - there's a password that should be removed to a local config file.

Two other notes:
- I only skimmed the unit tests, since I was trying to get these comments back to you as soon as I could. Let me know if you want me to take a look at them too
- There's a bunch of models getting loaded - they're handled well in the code, but do you have an estimate about how many system resources this will take? If not, we can deploy to dev backoffice and go from there.

golnazads commented:
@kelockhart 
Thank you for your comments. But you know, some of the issues you bring up, I don't know the answer to, for example, how often to update the synonyms file. So I am going to compile these and make it an issue for you/Alberto and anyone else in the know to answer them.

golnazads commented:
You said you only checked out the unittests. If you have time or inclination to check it out more, I am not going to have time to work on this until second week of August. So feel free to look some more and post feedback. There are two issues that I need a bit more explanation on, and two issues that we need to chat about since I dont know the answer. Thank you for check this out.

golnazads committed:
moved Brain password/url to config file

golnazads committed:
needed newer version of adsputil

golnazads committed:
replaced print with log in run.py, also latest celery needs latest version of python-dateutil, fixed that.

golnazads committed:
added function description/param explanation

golnazads committed:
added more comments to tasks, a bit of code moved around, and hence needed one more unittest to keep it at 100 percent coverage

golnazads committed:
added function description/param explanations

golnazads committed:
added function description/param explanation, also added a method to NamedEntityLabel class to sort the two items and alway have planetry listed first

golnazads committed:
fixed the check for return number of rows

golnazads committed:
per Taylor use a specific pip version to fix the issue with celery

golnazads committed:
added function description/param explanation

golnazads committed:
added function description/param explanations

golnazads committed:
added function description/param explanations

golnazads committed:
fixed the missing import

golnazads committed:
added couple of minor comments

golnazads committed:
added timestamp to EntityArgs

golnazads committed:
added function description, moved top astronomy list to config, and made timestamp be an arg passed in

golnazads committed:
moved default timestamp to config

golnazads committed:
added function description/param explanations

golnazads committed:
added function description/param explanations, also moved wiki-vocab to a dat file

golnazads commented:
@kelockhart 
OK, addressed all the comments. Please check it out and if OK merge it.
I will start working on populating the database as soon as @tjacovich give me the address to the database.
Would you create the deployed ymls?

kelockhart reviewed (approved):
Looks great! I have one small comment, but it doesn't affect anything now, so I'll make an issue to deal with for later.
